add `Base.length(::SkipMissing)` for `AbstractArray`

thazhemadam opened the pull request:
Add `Base.length` definition for `SkipMissing`.

Resolves #42672.

Before this change:
```julia
julia> x = [missing, 1, missing];

julia> (length∘skipmissing)(x)
ERROR: MethodError: no method matching length(::Base.SkipMissing{Vector{Union{Missing, Int64}}})
```

With this change:
```julia
julia> x = [missing, 1, missing];

julia> (length∘skipmissing)(x)
1
```

thazhemadam committed:
missing: add `length(::SkipMissing)`

Define `length(::SkipMissing)` with `O(N)` complexity.

bkamins commented:
Maybe we could restrict the signature to `skipmissing` on `AbstractArray`. This method is really needed (otherwise there is no easy way to check the number of elements of passed object that is `skipmissing`).

thazhemadam committed:
missing: restrict `length(::SkipMissing)` to only `AbstractArray`

`length(::SkipMissing)`, as defined, can actually change the length
of some objects (like stateful iterators [1]).

Thus, restrict the method to work only on `AbstractArray`.

[1] `length(::SkipMissing)` changing the length of a `Stateful` iterator.
```
julia> itr = Iterators.Stateful([missing, 1, missing])
Base.Iterators.Stateful{Vector{Union{Missing, Int64}}, Union{Nothing, Tuple{Union{Missing, Int64}, Int64}}}(Union{Missing, Int64}[missing, 1, missing], (missing, 2), 0)

julia> count(!ismissing, skipmissing(itr).x) # proposed definition of length
1

julia> count(!ismissing, skipmissing(itr).x)
0
```

bkamins reviewed (approved):
Looks good to me.

@thazhemadam - when this PR gets merged could you also open a follow-up PR to Compat.jl to ensure backward compatibility? Thank you!

thazhemadam commented:
>  @thazhemadam - when this PR gets merged could you also open a follow-up PR to Compat.jl to ensure backward compatibility? Thank you!

Yes, I can do that. 🙂

bkamins commented on test/missing.jl:
```diff
@@ -441,6 +441,17 @@ end
+        @test (length∘skipmissing)(somemiss) == 10
+
+        nomiss = rand(1:10, 10)
+        @test (length∘skipmissing)(nomiss) == 10
```
one small thing - maybe add a test for the case that fails because `skipmissing` wraps some iterator that is not an array?

thazhemadam replied:
Addressed in 5db9504f5c32ac207b8cdb1ce725dc7e46c1920f. Does this look good?

bkamins replied:
OK

thazhemadam committed:
test for error when length(::SkipMissing) is used with Stateful iterator

Add test checking if a `MethodError` is thrown while trying to use
`length(::SkipMissing)` with a Stateful iterator.

bkamins commented:
I think `skipmissing` should be an allowed exception. The reason is that, although in this case the operation is not O(1), in it very often used in statistics and the fact that `length` does not work on it is a major usability issue reported by many users.

nalimilan commented:
> Contractually, length should probably be O(1) though; and we hope to fix that for String at some point.

Is that really possible? Note that we also define `length(graphemes(::String))`, which is unlikely to be made O(1).

Marking for triage to get a decision.

JeffBezanson commented:
This feels weird to me but it's hard to point to a real problem it causes, so might as well?
